lib/src/rules/prefer_final_fields.dart

```diff
@@ -135,21 +152,44 @@ class _Visitor extends SimpleAstVisitor<void> {
       return;
     }
 
-    fields.variables.forEach((VariableDeclaration variable) {
+    for (var variable in fields.variables) {
       final element = variable.declaredElement;
-      if (!element.isPrivate) {
-        return;
-      }
-
-      if (variable.initializer == null) {
-        return;
-      }
 
-      if (_mutatedFields.contains(element)) {
-        return;
+      if (element.isPrivate && !_mutatedFields.contains(element)) {
+        bool fieldInConstructor(constructor) =>
+            constructor.initializers.any((initializer) =>
+                _containedInInitializer(element, initializer)) ||
+            constructor.parameters.parameters
+                .any((formal) => _containedInFormal(element, formal));
+
+        final classDeclaration = node.parent;
+        final constructors = (classDeclaration is ClassDeclaration
+            ? classDeclaration.members
+                .whereType<ConstructorDeclaration>()
+                .toSet()
+            : <ConstructorDeclaration>{});
+        final isFieldInConstructors = constructors.any(fieldInConstructor);
+        final isFieldInAllConstructors = constructors.every(fieldInConstructor);
+
+        if (isFieldInConstructors) {
+          if (isFieldInAllConstructors) {
+            rule.reportLint(variable);
+          }
+        } else if (element.initializer != null) {
+          rule.reportLint(variable);
+        }
       }
-
-      rule.reportLint(variable);
-    });
+    }
   }
 }
+
+bool _containedInInitializer(
+        Element element, ConstructorInitializer initializer) =>
+    initializer is ConstructorFieldInitializer &&
+    DartTypeUtilities.getCanonicalElementFromIdentifier(
+            initializer.fieldName) ==
+        element;
+
+bool _containedInFormal(Element element, FormalParameter formal) =>
+    formal is FieldFormalParameter &&
+    formal.declaredElement.toString() == element.toString();
```

Comment:
It ought to work to compare the elements without converting them to strings first and it ought to be safer in that we can't end up with two different elements that happen to produce the same string incorrectly comparing to be equal.